syntax = "proto3";
// 类似命名空间
package my_chat;

option cc_generic_services = true;

//用户信息结构
message UserInfo {
    string user_id = 1;//用户ID
    string nick_name = 2;//昵称
    string person_signature = 3;//个人签名/描述
    string phone_number = 4; //绑定手机号
    bytes  head_portrait = 5;//头像照片，文件内容使用二进制
}

//聊天会话信息
message ChatSessionInfo {
    optional string single_chat_friend_id = 1;//群聊会话不需要设置，单聊会话设置为对方ID
    string chat_session_id = 2; //会话ID
    string chat_session_name = 3;//会话名称git 
    optional MessageInfo prev_message = 4;//会话上一条消息，新建的会话没有最新消息
    optional bytes head_portrait = 5;//会话头像 --群聊会话不需要，直接由前端固定渲染，单聊就是对方的头像
}

//消息类型
enum MessageType {
    TEXT = 0;
    IMAGE = 1;
    FILE = 2;
    VOICE = 3;
}
message TextMessageInfo {
    string text_content = 1;//文字聊天内容
}
message ImageMessageInfo {
    optional string file_id = 1;//图片文件id,客户端发送的时候不用设置，由transmit服务器进行设置后交给storage的时候设置
    optional bytes image_content = 2;//图片数据，在ES中存储消息的时候只要id不要文件数据, 服务端转发的时候需要原样转发
}
message FileMessageInfo {
    optional string file_id = 1;//文件id,客户端发送的时候不用设置
    int64 file_size = 2;//文件大小
    string file_name = 3;//文件名称
    optional bytes file_contents = 4;//文件数据，在ES中存储消息的时候只要id和元信息，不要文件数据, 服务端转发的时候也不需要填充
}
message VoiceMessageInfo {
    optional string file_id = 1;//语音文件id,客户端发送的时候不用设置
    optional bytes voice_contents = 2;//文件数据，在ES中存储消息的时候只要id不要文件数据, 服务端转发的时候也不需要填充
}
message MessageContent {
    MessageType message_type = 1; //消息类型
    oneof msg_content {
        TextMessageInfo text_message = 2;//文字消息
        FileMessageInfo file_message = 3;//文件消息
        VoiceMessageInfo voice_message = 4;//语音消息
        ImageMessageInfo image_message = 5;//图片消息
    };
}
//消息结构
message MessageInfo {
    string message_id = 1;//消息ID
    string chat_session_id = 2;//消息所属聊天会话ID
    int64 timestamp = 3;//消息产生时间
    UserInfo sender = 4;//消息发送者信息
    MessageContent message = 5;
}

message FileDownloadData {
    string file_id = 1;
    bytes file_content = 2;
}

message FileUploadData {
    string file_name = 1;
    int64 file_size = 2;
    bytes file_content = 3;
}